app/controllers/base_metric_configurations_controller.rb

```diff
@@ -1,42 +1,147 @@
+
+    respond_to do |format|
+      format.html { redirect_to kalibro_configuration_path(@kalibro_configuration.id) }
+      format.json { head :no_content }
+    end
   end
 
   protected
 
-  def metric_configuration
-    raise NotImplementedError
+  def save_and_redir(format)
+    new_record = @metric_configuration.id.nil?
+    result = block_given? ? (yield @metric_configuration) : @metric_configuration.save
+
+    if result
+      clear_caches
+
+      format.html do
+        redirect_to kalibro_configuration_path(@kalibro_configuration.id),
+          notice: t(new_record ? 'successfully_created' : 'successfully_updated',
+                    record: t(@metric_configuration.class))
+      end
+      format.json { render json: @metric_configuration, status: new_record ? :created : :ok }
+    else
+      failed_action(format)
+    end
   end
 
-  def update_metric_configuration (new_metric_configuration)
-    raise NotImplementedError
+  def failed_action(format, error = nil)
+    errors = @metric_configuration.kalibro_errors
+    errors << error unless error.nil?
+
+    flash[:notice] = errors.join(', ')
+
+    format.json { render json: { errors: errors }, status: :unprocessable_entity }
+    render_failure_html(format)
   end
 
-  # Never trust parameters from the scary internet, only allow the white list through.
-  # TODO: this should be refactored to the concern metric configuration
+  def render_failure_html(format)
+    if action_name == 'create'
+      format.html { render 'new' }
+    elsif action_name == 'update'
+      format.html { render 'edit' }
+    else
+      format.html { redirect_to kalibro_configuration_path(@kalibro_configuration.id) }
+    end
+  end
+
+  def clear_caches
+    Rails.cache.delete("#{@kalibro_configuration.id}_tree_metric_configurations")
+    Rails.cache.delete("#{@kalibro_configuration.id}_hotspot_metric_configurations")
+  end
+
+  # Notice: If you add some logic to this method, remove the :nocov: below
+  # :nocov:
   def metric_configuration_params
-    params[:metric_configuration]
+    raise NotImplementedError
+  end
+  # :nocov:
+
+  def set_kalibro_configuration!
+    @kalibro_configuration = KalibroConfiguration.find params[:kalibro_configuration_id].to_i
+  end
+
+  def new_metric_configuration!
+    @metric_configuration = MetricConfiguration.new metric_configuration_params
+    @metric_configuration.kalibro_configuration_id = @kalibro_configuration.id
+  end
+
+  def find_metric_configuration!
+    @metric_configuration = MetricConfiguration.find params[:id].to_i
+
+    # Make sure the metric configuration is really from the kalibro configuration we're being told it is
+    if @metric_configuration.kalibro_configuration_id != @kalibro_configuration.id
+      raise KalibroClient::Errors::RecordNotFound
+    end
+  end
+
+  def set_metric!
+    collector = KalibroClient::Entities::Processor::MetricCollectorDetails.find_by_name(params[:metric_collector_name])
+    unless collector.nil?
```

Comment:
Seems like a good idea. The only issue is that returning a 404 is not ideal in this case, since the URL is actually valid, but one of the sent parameters is not.